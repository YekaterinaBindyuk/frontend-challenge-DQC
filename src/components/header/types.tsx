import type { Question } from '../shared/types'

export interface HeaderProps {
  title: string
  questions: Question[]
  creationDate: string
}
