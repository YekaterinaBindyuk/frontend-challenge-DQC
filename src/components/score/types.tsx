import type { ScaleQuestion } from '../shared/types'

export interface HappinessScoreProps {
  questions: ScaleQuestion[]
}
