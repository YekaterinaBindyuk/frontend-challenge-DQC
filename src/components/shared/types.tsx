export enum QuestionType {
  Text = 'text',
  Number = 'number',
};

export interface Question {
  question_text: string
  type: QuestionType
  responses: Array<number | string>;
}

export interface ScaleQuestion extends Question {
  type: QuestionType.Number
  responses: number[]
}

export interface FreeTextQuestion extends Question {
  type: QuestionType.Text
  responses: string[]
}

export interface SurveyResults {
  survey_title: string
  created_at: string
  questions: Question[] 
}

const isQuestionOfType = <T extends QuestionType>(type: T) =>
    (question: Question): question is Extract<Question, { type: T }> =>
        question.type === type;

export const isFreeTextQuestion = isQuestionOfType(QuestionType.Text);
export const isScaleQuestion = isQuestionOfType(QuestionType.Number);