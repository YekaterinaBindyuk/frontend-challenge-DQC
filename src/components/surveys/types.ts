import { type FreeTextQuestion } from '../shared/types'

export interface SurveyFreeTextProps {
  freeTextQuestions: FreeTextQuestion[]
}
