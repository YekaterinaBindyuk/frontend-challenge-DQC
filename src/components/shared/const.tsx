export const MAX_HAPPINESS_SCORE = 100
export const HAPPINESS_SCALE = 5
export const PREFERABLE_LOCALE = 'de-DE'
