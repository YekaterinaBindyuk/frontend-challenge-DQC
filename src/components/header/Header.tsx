import type { FunctionComponent } from 'react'

import { FontIcon } from '@fluentui/react'

import { getFormattedDate, getMaxResponsesCount } from './helpers'
import type {  HeaderProps } from './types'
import { HappinessScore } from '../score/HappinessScore'
import { isScaleQuestion, type ScaleQuestion } from '../shared/types'

import styles from './Header.module.css'

export const Header: FunctionComponent<HeaderProps> =
 ({ title, questions, creationDate }) => {
   const creationDay = getFormattedDate(creationDate)
   const respondersCount = getMaxResponsesCount(questions)
   const scaleQuestions: ScaleQuestion[] =
   questions.filter<ScaleQuestion>( isScaleQuestion)
   
   return (<><h1>
        <FontIcon iconName="ClipboardList" className={styles.icon} />
        {title}
        </h1>
        <div>
        <span data-testid="creationDay">
          This survey was started on {creationDay}.
        </span>
       {/* eslint-disable-next-line max-len */}
        <span data-testid="respondersCount"> Overall, {respondersCount} people participated in the survey.
        </span>
        </div>
        <HappinessScore questions={scaleQuestions}/>
        </>)
 }
