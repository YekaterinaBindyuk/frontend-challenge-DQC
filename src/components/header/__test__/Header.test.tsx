import { render, type RenderResult, screen } from '@testing-library/react'

import { QuestionType } from '../../shared/types'
import { Header } from '../Header'

const mockedQuestionsData = [{
  question_text: 'Which skills would you like to learn?',
  type: QuestionType.Text,
  responses: [
    'Data quality evaluation',
    'Analytics in BI tools',
    'Data visualization',
    'Analytics',
    'None'
  ]
}
]

describe('Header Component', () => {
  const setup = (): RenderResult => {
    return render(<Header
            title='business data'
            questions={mockedQuestionsData}
            creationDate='2022-11-16T15:34:53'
        />)
  }
  test('renders the component', () => {
    const container = render(<Header
      title='business data'
      questions={mockedQuestionsData}
      creationDate='2022-11-16T15:34:53'
  />)
    expect(container).toMatchSnapshot()
  })
  test('should render Header FC', () => {
    setup()
    const title = screen.getByText(/business data/i)
    expect(title).toBeInTheDocument()
  })

  test('should render amount of responses', () => {
    setup()
    const respondersSpan = screen.getByTestId('respondersCount')
    expect(respondersSpan).toHaveTextContent(/5/)
  })

  test('should render creation date', () => {
    setup()
    const creationDaySpan = screen.getByTestId('creationDay')
    expect(creationDaySpan).toHaveTextContent(/16.11.2022/)
  })
})
