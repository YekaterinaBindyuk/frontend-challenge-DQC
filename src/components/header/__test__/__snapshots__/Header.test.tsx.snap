// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        <i
          aria-hidden="true"
          class="ms-Icon root-105 icon"
          data-icon-name="ClipboardList"
        />
        business data
      </h1>
      <div>
        <span
          data-testid="creationDay"
        >
          This survey was started on 
          16.11.2022
          .
        </span>
        <span
          data-testid="respondersCount"
        >
           Overall, 
          5
           people participated in the survey.
        </span>
      </div>
      <h1
        data-testid="happinessScore"
      >
        <i
          aria-hidden="true"
          class="ms-Icon root-105 icon"
          data-icon-name="ChatBot"
        />
        0
         / 
        100
      </h1>
    </div>
  </body>,
  "container": <div>
    <h1>
      <i
        aria-hidden="true"
        class="ms-Icon root-105 icon"
        data-icon-name="ClipboardList"
      />
      business data
    </h1>
    <div>
      <span
        data-testid="creationDay"
      >
        This survey was started on 
        16.11.2022
        .
      </span>
      <span
        data-testid="respondersCount"
      >
         Overall, 
        5
         people participated in the survey.
      </span>
    </div>
    <h1
      data-testid="happinessScore"
    >
      <i
        aria-hidden="true"
        class="ms-Icon root-105 icon"
        data-icon-name="ChatBot"
      />
      0
       / 
      100
    </h1>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
